src/core/namespace.c

```diff
@@ -3291,6 +3309,518 @@ int open_shareable_ns_path(int ns_storage_socket[static 2], const char *path, un
+
+        if (fsconfig(overlay_fs_fd, FSCONFIG_SET_STRING, "lowerdir", options, 0) < 0)
+                return log_debug_errno(errno, "Failed to set lowerdir=%s: %m", options);
+
+        if (fsconfig(overlay_fs_fd, FSCONFIG_SET_STRING, "source", "systemd-extensions", 0) < 0)
+                return log_debug_errno(errno, "Failed to set source=systemd-extensions: %m");
+
+        /* Create the superblock */
+        if (fsconfig(overlay_fs_fd, FSCONFIG_CMD_CREATE, NULL, NULL, 0) < 0)
+                return log_debug_errno(errno, "Failed to create overlay superblock: %m");
+
+        /* Signal completion to grandchild */
+        r = send_one_fd(pipe_fd, overlay_fs_fd, 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to signal overlay configuration complete for %s: %m", mount_path);
+
+        return 0;
+}
+
+static int refresh_apply_and_prune(const NamespaceParameters *p, MountList *ml) {
+        int r;
+
+        assert(p);
+        assert(ml);
+
+        /* Open all extensions on the host, drop all sysexts since they won't have /etc/. The list of
+         * overlays also need to be updated, so that if it's empty after a confext has been removed, the
+         * child process can correctly undo the overlay in the target namespace, rather than attempting to
+         * mount an empty overlay which the kernel does not allow, so this pruning has to be done here and
+         * not later (nor earlier, as we don't know if an image is a confext until this point). */
+        MountEntry *f, *t;
+        for (f = ml->mounts, t = ml->mounts; f < ml->mounts + ml->n_mounts; f++) {
+                if (IN_SET(f->mode, MOUNT_EXTENSION_DIRECTORY, MOUNT_EXTENSION_IMAGE)) {
+                        f->filter_class = IMAGE_CONFEXT;
+
+                        r = apply_one_mount("/", f, p);
+                        if (r < 0)
+                                return r;
+                        /* Nothing happened? Then it is not a confext, prune it from the lists */
+                        if (r == 0) {
+                                FOREACH_ARRAY(m, ml->mounts, ml->n_mounts) {
+                                        if (m->mode != MOUNT_OVERLAY)
+                                                continue;
+
+                                        _cleanup_strv_free_ char **to_remove = NULL;
+
+                                        STRV_FOREACH(ol, m->overlay_layers)
+                                                if (startswith(*ol, mount_entry_path(f))) {
+                                                        r = strv_extend(&to_remove, *ol);
+                                                        if (r < 0)
+                                                                return log_oom_debug();
+                                                }
+                                        STRV_FOREACH(ol, to_remove)
+                                                m->overlay_layers = strv_remove(m->overlay_layers, *ol);
```

Comment:
hmm, creating a list to remove its contents? that's a bit weird, no?

just use strv_filter_prefix()?

Reply:
extend strv_filter_prefix() to take a boolean that negates the check?